### PR TITLE
US70399 - add property to turn off use of subdomains to get zone

### PR DESCRIPTION
**Pull request by FrankGasparovic**

### CODE REVIEW

**Comment by FrankGasparovic** [comment]
Passed 
https://predix1.jenkins.build.ge.com/job/Predix-Security/token-lib/token-lib-dev-pullrequest/115//badge/icon

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 1]
```diff
@@ -57,45 +57,62 @@
     public void testLoadAuthentication() {
         // testing when zone id is not null
         String zoneUserScope = SERVICEID + ".zones." + ZONE + ".user";
-        assertAuthentication(loadAuthentication(PREDIX_ZONE_HEADER_NAME, BASE_DOMAIN, ZONE, zoneUserScope), ZONE);
+        assertAuthentication(loadAuthentication(PREDIX_ZONE_HEADER_NAME, BASE_DOMAIN, ZONE, zoneUserScope, true), ZONE);
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: Not reviewing this class for now.    Wasnt expecting so many changes to this class for this change.

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 2]
```diff
@@ -210,17 +224,63 @@ private HttpServletRequest mockHttpRequest(String requestZoneName, String reques
...
+            List<String> nonZoneUriPatterns, HttpServletRequest request) {
 
         ZacTokenService zacTokenServices = new ZacTokenService();
+        RestTemplate mockRestTemplate = configureMockRestTemplate();
+        DefaultZoneConfiguration zoneConfig = new DefaultZoneConfiguration();
+
+        zoneConfig.setTrustedIssuerIds(trustedIssuers);
+        zoneConfig.setAllowedUriPatterns(nonZoneUriPatterns);
+
+        zacTokenServices.setDefaultZoneConfig(zoneConfig);
```
Comment (sanjeevchopra) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: is default zone config being used by any test ? if not, pls remove.

Comment (FrankGasparovic) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: It is being used

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ public void testGetZoneNameWithHeaderAndBaseDomains(final String requestHostname
         MockHttpServletRequest req = new MockHttpServletRequest();
         req.setServerName(requestHostname);
         req.addHeader(requestHeader, requestHeaderValue);
-        String actualZone = HttpServletRequestUtil.getZoneName(req, serviceBaseDomains, serviceConfigHeaders);
+        String actualZone = HttpServletRequestUtil.getZoneName(req, serviceBaseDomains, serviceConfigHeaders, true);
```
Comment (FrankGasparovic) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: Added

File: src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java [hunk 4]
```diff
@@ -236,4 +238,8 @@ public void setDefaultZoneConfig(final DefaultZoneConfiguration defaultZoneConfi
         return this.serviceBaseDomainList;
     }
 
+    public void setEnableSubdomainsForZones(final Boolean useSubDomainsForZones) {
+        this.enableSubdomainsForZones = useSubDomainsForZones;
```
Comment (FrankGasparovic) [inline src/main/java/com/ge/predix/uaa/token/lib/AbstractZoneAwareTokenService.java]: Changed

File: src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java [hunk 5]
```diff
@@ -143,4 +153,18 @@ public void testGetZoneNameWithHeaderAndBaseDomains(final String requestHostname
...
+                { "acs.com", Arrays.asList("acs.com", "guardians.com"),
+                            Collections.singletonList("Predix-Zone-Id"), null, false },
+                { "zone1.acs.com", Arrays.asList("acs.com", "guardians.com"),
+                        Collections.singletonList("Predix-Zone-Id"), "zone1", true },
+                { "acs.com", Arrays.asList("acs.com", "guardians.com"),
+                            Collections.singletonList("Predix-Zone-Id"), null, true }};
+
+    }
+
 }
```
Comment (irinaepshteyn) [inline src/test/java/com/ge/predix/uaa/token/lib/HttpServletRequestUtilTest.java]: @FrankGasparovic, 
Missing test use cases:
- enableZonesAsSubdomains=false and zone is passed in a header and as a subdomain.  Should return zone from the header
- enableZonesAsSubdomains = false and zone is passed in a header.  Should return zone from the header
- enableZonesAsSubdomains = true and zone is passed in a header.  Should return zone from the header

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 6]
```diff
@@ -143,55 +160,52 @@ public void testDefaultAndZoneSpecificResourceAuthorization(final String zoneId,
     }
 
     private OAuth2Authentication loadAuthentication(String configuredHeaderNames, String configuredBaseDomains,
-            final String requestZoneName, final String userScopes) {
-        return loadAuthentication(configuredHeaderNames, configuredBaseDomains, requestZoneName, userScopes,
-                "/test/resource", Arrays.asList("/zone/**"));
+            final String requestZoneName, final String userScopes, Boolean useSubdomainsForZones) {
+        return loadAuthenticationWithZoneAsHeader(configuredHeaderNames, configuredBaseDomains, requestZoneName,
+                useSubdomainsForZones, userScopes, "/test/resource", Arrays.asList("/zone/**"));
```
Comment (irinaepshteyn) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: To improve readability, remove loadAuthentication and call loadAuthenticationWithZoneAsHeader explicitly.

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 7]
```diff
@@ -57,45 +57,62 @@
...
         // testing when zone id is not null
-        ZacTokenService zacTokenServices = configureZacTokenService("", null, "");
+        ZacTokenService zacTokenServices = configureZacTokenService("", null, "", true, Collections.EMPTY_LIST,
+                Collections.EMPTY_LIST, null);
         zacTokenServices.afterPropertiesSet();
     }
 
+    @Test(expectedExceptions = InvalidRequestException.class)
+    public void testLoadAuthenticationNoSubdomains() throws Exception {
+        // testing when zone id is not null
```
Comment (irinaepshteyn) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: testLoadAuthenticationNoSubdomains duplicates testLoadAuthenticationUnauthorizedScope.  The only difference is how the zone is passed in which is covered by tests for HttpServletRequestUtil.getZoneName

File: src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java [hunk 8]
```diff
@@ -57,45 +57,62 @@
...
+        String zoneUserScope = "scope.not.used";
+        loadAuthenticationWithZoneAsSubdomain(PREDIX_ZONE_HEADER_NAME, BASE_DOMAIN, ZONE, false, zoneUserScope,
+                "/test/resource", Arrays.asList("/zone/**"));
+    }
+
+    public void testLoadAuthenticationWithSubdomains() throws Exception {
+        // testing when zone id is not null
```
Comment (irinaepshteyn) [inline src/test/java/com/ge/predix/uaa/token/lib/ZacTokenServiceTest.java]: testLoadAuthenticationWithSubdomains duplicates testLoadAuthentication.  The only difference is how the zone is passed in which is covered by tests for HttpServletRequestUtil.getZoneName

**Pull request merged**